### PR TITLE
Light Client Bridge Stubbed Implementation

### PR DESCRIPTION
**Pull request by musnit**
This PR follows from https://github.com/Snowfork/verification-solidity/pull/7

It adds a few comments and pieces of pseudocode for missing steps in the LightClientBridge, as well as some comments in the test for the lightClientBridge for suggested additional test cases.

@hcaw is currently working on fixes to the ValidatorRegistry contract + tests

@philipstanislaus is going to look at the LightClientBridge and this epic as a whole and take some ownership of leading completion of the work. Suggested plan is:

 - Review and complete the pseudocode/stubbed out code, making sure that interfaces, input/output variables are defined properly and that steps are reviewed and written in pseudocode either within the LightClientBridge or possibly abstracted out into other library/tooling contracts
 - Complete working implementation with happy-path tests working of the interactive protocol newSignatureCommitment + completeSignatureCommitment flow
 - Integrate relayer with contract call endpoints and get e2e relaying working for happy path, with a lot of steps/internal verification actually being skipped, and with empty input data for the proofs if they are not available yet.
 - Extend the test cases to cover tests for all sad paths, attack vectors and edge cases. Some suggested test cases to be added are already in comments at the bottom of lightClientBridge.ts that can be implemented for failing tests
 - Start completing implementation of stubbed out/pseudocode internal verification steps, getting above test cases to slowly pass, including completing implementation of other libraries used like ValidatorRegistry tests _(@hcaw currently on this)_
 - Complete implementation of relayer, ensuring it does infact get the correct MMR/Merkle proofs and submit them all as needed to the contract.
 - Complete implementation of statement processing and of applying validator set changes
 - Ensure mmr-per-epoch requirements and epoch-changes are being handled correctly
 - Get all tests passing and contract completed

### CODE REVIEW

File: contracts/LightClientBridge.sol [hunk 1]
```diff
@@ -116,25 +124,19 @@ contract LightClientBridge {
      * @param statement contains the statement signed by the validator(s)
      * @param randomSignatureCommitments an array of signatures from the randomly chosen validators
      * @param randomSignatureBitfieldPositions an array of bitfields from the chosen validators
```
Comment (philipstanislaus) [inline contracts/LightClientBridge.sol:126]: It would be much more efficient to just submit the signature bitfield instead. This consumes N bytes for N signers, while a bitfield consumes M bits for M validators. The bitfield is 8x more efficient in the best case.

**Review by philipstanislaus — COMMENTED** [review]
Thanks 👍 

Couple of open points, please have a look @musnit and @hcaw. I'll also push a commit with pending tests and a few TODOs.

File: contracts/LightClientBridge.sol [hunk 2]
```diff
@@ -96,11 +96,16 @@ contract LightClientBridge {
             "Error: Sender must be in validator set"
         );
 
+        // TODO also add a check here that the signing validator is in validatorClaimsBitfield
```
Comment (philipstanislaus) [inline contracts/LightClientBridge.sol:99]: That check is missing currently, added this TODO^^

File: contracts/LightClientBridge.sol [hunk 3]
```diff
@@ -137,6 +142,8 @@ contract LightClientBridge {
     ) public {
         ValidationData storage data = validationData[id];
 
+        // TODO verify that sender is the same as in `newSignatureCommitment` (who? singer or relayer?)
```
Comment (philipstanislaus) [inline contracts/LightClientBridge.sol:145]: That check is missing currently, added this TODO^^

File: contracts/LightClientBridge.sol [hunk 4]
```diff
@@ -160,7 +167,7 @@ contract LightClientBridge {
             //  valid based on the ValidatorRegistry merkle root, ie, confirm that
             //  the randomSignerAddress is from an active validator and is at the correct position
             require(
-                validatorRegistry.checkValidatorInSet(randomValidatorAddresses[i], randomPublicKeyMerkleProofs[i], i),
+                validatorRegistry.checkValidatorInSet(randomValidatorAddresses[i], randomPublicKeyMerkleProofs[i]),
```
Comment (philipstanislaus) [inline contracts/LightClientBridge.sol:170]: `i` needs to be verified here as well, I just removed it because the test would fail for me. Please add it again and change the implementation to check the position in the merkle tree.

File: contracts/LightClientBridge.sol [hunk 5]
```diff
@@ -213,14 +220,14 @@ contract LightClientBridge {
             // @note Type conversion from bytes32 -> uint8, by way of bytes1 (to work around limitations)
             onChainRandNums[i] = uint8(bytes1(randomSeedBlockHash));
         }
+        // TODO this might lead to duplicate entries?
```
Comment (philipstanislaus) [inline contracts/LightClientBridge.sol:223]: The random number generator above could yield duplicate positions. As mentioned elsewhere, it's also quite inefficient not working with a bitfield here. 

I'd actually propose to just hash some data into a word, and then add/remove bits until the right number of bits are set.

File: contracts/MerkleTree.sol [hunk 6]
```diff
@@ -41,5 +41,7 @@ contract MerkleTree {
      * @notice Verify that a specific leaf element is part of the Merkle Tree at a specific position in the tree
      * @return a boolean value representing the success or failure of the operation
      */
-    function verifyAtPosition()
+    // function verifyAtPosition() {
+    //     // TODO
+    // }
```
Comment (philipstanislaus) [inline contracts/MerkleTree.sol:46]: This is missing and should be added, just commenting it out here to make the tests run.

File: test/lightClientBridge.ts [hunk 7]
```diff
@@ -80,13 +80,24 @@ describe.only("LightClientBridge Contract", function () {
...
         senderSignatureCommitment,
+        signer.address,
         senderPublicKeyMerkleProof
       )
       expect(result).to.not.be.reverted
+
+      expect(await lightClientBridgeContract.count()).to.equal(1)
+
+      // TODO add assertion for the stake being locked up (whose stake? signer? or relayer?)
+      // TODO add assertion for the event being emitted
```
Comment (philipstanislaus) [inline test/lightClientBridge.ts:91]: Please add these two assertions

File: test/lightClientBridge.ts [hunk 8]
```diff
@@ -80,13 +80,24 @@ describe.only("LightClientBridge Contract", function () {
...
     })
+
+    it("Should revert when validatorPublicKey is not in in validatorRegistry given validatorPublicKeyMerkleProof")
+    it("Should revert when validatorPublicKey is not in validatorClaimsBitfield")
+    it("Should revert when validatorPublicKey is not signer of statement in validatorSignatureCommitment")
+    it("Should revert when validatorClaimsBitfield is too short")
```
Comment (philipstanislaus) [inline test/lightClientBridge.ts:97]: We should add these tests

File: test/lightClientBridge.ts [hunk 9]
```diff
@@ -127,37 +139,33 @@ describe.only("LightClientBridge Contract", function () {
...
+        randomSignatureBitfieldPositions as any,
+        randomSignerAddresses as any,
+        randomPublicKeyMerkleProofs as any
       )
+
+      // TODO add assertion for the stake being refundend
+
+      // TODO add assertion for processStatement being called
+
+      // TODO add assertion for event being emitted
```
Comment (philipstanislaus) [inline test/lightClientBridge.ts:152]: Again, please add these assertions

File: test/lightClientBridge.ts [hunk 10]
```diff
@@ -127,37 +139,33 @@ describe.only("LightClientBridge Contract", function () {
...
+    it("Should revert when a randomValidatorAddresses is not signer of statement in randomSignatureCommitments")
+    it("Should revert when statement is not for the current validator set") // TODO is that not covered above? Do we need other test cases here?
+    it("Should revert when statement is older than latest committed one") // TODO is that not covered above? Do we need other test cases here?
+    it("Should revert when statement is not for a previous epoch") // TODO is that not covered above? Do we need other test cases here?
+    it("Should revert when statement is not for the next epoch but has no validator set changes") // TODO is that not covered above? Do we need other test cases here?
+    it("Should revert when statement is not for a future epoch after the next") // TODO is that not covered above? Do we need other test cases here?
+
+    context("validator set changes", function () {
+      it("should correctly call the method to update the validator set of mockValidatorRegistry")
     })
```
Comment (philipstanislaus) [inline test/lightClientBridge.ts:169]: And a bunch more tests to be added

File: test/lightClientBridge.ts [hunk 11]
```diff
@@ -127,37 +139,33 @@ describe.only("LightClientBridge Contract", function () {
...
-  - processing:
-    - correct statement, but actually an old one or one without latest new mmr
-    - correct statement, but skips some epochs so opens a gap?
-    - correct statement, but doesnt actually include correct mmrproof to newest leaf
-    - correct statement, but actually includes mmrproof to old leaf
-    - correct statement, correct mmrproof, but actually wrong beefy_authority_set_merkle_root
-
-- can the Bits.sol implementation be games/is it fully secure? assess how it works in detail and think about attack vectors on it.
-
-*/
```
Comment (philipstanislaus) [inline test/lightClientBridge.ts]: These have all been added as pending tests or TODOs above.

File: contracts/LightClientBridge.sol (hunk 3 above)
Comment (philipstanislaus) [reply contracts/LightClientBridge.sol:145]: Should be the relayer, check with `msg.sender`

File: contracts/LightClientBridge.sol [hunk 12]
```diff
@@ -224,34 +229,34 @@ contract LightClientBridge {
...
+    function processPayload(bytes32 payload) private {
+        // Check the payload is newer than the latest
+        // Check that payload.leaf.block_number is > last_known_block_number;
 
-        // Check the statement is within the last epoch, or for the next epoch
-        // Todo
+        //update latestMMRRoot = payload.mmrRoot;
 
-        // latestMMRRoot = statement.mmrRoot;
+        // if payload is in next epoch, then apply validatorset changes
```
Comment (philipstanislaus) [inline contracts/LightClientBridge.sol:241]: We  also need to reject if for later than next epoch

**Review by philipstanislaus — COMMENTED** [review]
LGTM, jus one minor comment

**Comment by philipstanislaus** [comment]
Merged indirectly into `master` through #16
